[AIRFLOW-276] Refresh stale dags

ldct opened the pull request:
Dear Airflow Maintainers,

Please accept this PR that addresses the following issues:
- https://issues.apache.org/jira/browse/AIRFLOW-276

Parsing and executing dag files can be slow, since they are python scripts. Hence, we cannot reload the dags folder on every request. The current code deals with this by only loading the dags folder once on startup, so it doesn't pick up new changes.

This PR starts, for each flask (request-serving) worker process, a thread that continuously loads the dags folder. Even if that is slow, the main thread will still serve requests quickly. The threads share a single dagbag object, and once the child thread finishes loading the dags folder, it updates the dagbag object.

There are a few alternative strategies:

Option 1. We could call `refresh_dags` on every page view. Most of the time the file timestamps will be unchanged and this will be fast. However if an expensive-to-load dag is added, an unlucky user will have to wait a few seconds for his request to be served.
Option 2. We could make the webserver code not load the dags folder at all, and read exclusively from the airflow db. 

Concretely, this option means
1. A new dag is uploaded to the scheduler's dag folder, or an existing dag is updated
2. The scheduler loads the new dag
3. The scheduler partially serialized the new dag and places it in the DB (it cannot be fully serialized since dags are not in general pickleable)
4. The web UI only renders from the db (it does not need to have the .py file), and so picks up the change

Disadvantages of option 2:
1. Dags must know how to partially serialize themselves in sufficient detail that the web UI can display all the existing details about them
2. More data must be put into the db than is currently there
3. If the scheduler is not up, the UI will not be updated.

Option 3. A separate process (instead of thread) could be started on the web UI, which passes a partially serialized dag to the web process. This also has the disadvantage that dags must know how to partially serialize themselves.
### Testing Done:
- [ ] start webserver, add a new dag to `~/airflow/dags`, make sure the new change is visible in the web UI
- [ ] add a new dag with a syntax error to `~/airflow/dags`, make sure the web UI is still up

@plypaul @mistercrunch

jlowin committed:
Handle queued tasks from multiple jobs/executors

When Scheduler is run with `—num-runs`, there can be multiple
Schedulers and Executors all trying to run tasks. For queued tasks,
Scheduler was previously only trying to run tasks that it itself had
queued — but that doesn’t work if the Scheduler is restarting. This PR
reverts that behavior and adds two types of “best effort” executions —
before running a TI, executors check if it is already running, and
before ending executors call sync() one last time

r39132 committed:
Merge pull request #1400 from jlowin/missing-args-clear

Add missing args to `airflow clear` - confirmed this change works locally.

bolkedebruin committed:
Revert from using "--foreground" to "--daemon"

This reinstalls old behavior or running on the foreground by default.

bolkedebruin committed:
Use os.execvp instead of subprocess.Popen for the webserver

subprocess.Popen forks before doing execv. This makes it difficult
for some manager daemons (like supervisord) to send kill signals.
This patch uses os.execve directly. os.execve takes over the current
process and thus responds correctly to signals

* Resolves residue in ISSUE-852

Commit:
Add columns to toggle extra detail in the connection list view.

johnnason committed:
Reinstate imports for github enterprise auth

Pylint based refactoring removed a necessary import. 

* Resolves #1303

bolkedebruin committed:
Merge pull request #1420 from bolkedebruin/foreground_default

Make foreground default and use os.execvp in specific cases

Commit:
correct missed arg.foreground to arg.daemon in cli

bolkedebruin committed:
Merge pull request #1426 from jgao54/bug-fix

correct missed arg.foreground to arg.daemon in cli

jlowin committed:
Clean up issue template (#1419)

mistercrunch committed:
Log dagbag metrics dupplicate messages in queue into Statsd (#1406)

plypaul committed:
Log the number of errors when importing DAGs

whummer committed:
enable UI feature to recursively set success=True for all operators within SubDagOperator

jlowin committed:
Merge pull request #1429 from plypaul/plypaul_log_import_errors

Log the number of errors when importing DAGs

r39132 committed:
Merge pull request #1422 from stranbird/conn-ui-revamp

Show `extra` data on the connection list view

mistercrunch committed:
Revert "Show `extra` data on the connection list view" (#1438)

r39132 committed:
Replace Github wiki links with Apache cwiki links

r39132 committed:
Merge pull request #1447 from r39132/master

Replace Github wiki links with Apache cwiki links

criccomini committed:
AIRFLOW-15: Remove gcloud

sharifyounes committed:
Update dags.html

criccomini committed:
Merge pull request #1448 from criccomini/remove-gcloud

AIRFLOW-15: Remove gcloud

r39132 committed:
Merge pull request #1449 from sharifyounes/master

Update dags.html

Commit:
Add Zendesk as a company which is using Airflow

r39132 committed:
Merge pull request #1450 from jwswj/jsmale/add_zendesk

Add Zendesk as a company using Airflow

criccomini committed:
AIRFLOW-16: Update Google cloud hooks to use new Google cloud platform UI.

criccomini committed:
Merge pull request #1452 from criccomini/improve-gcp-hooks

AIRFLOW-16: Update Google cloud hooks to use new Google cloud platfor…

alexvanboxel committed:
AIRFLOW-21 upgrade GCP client lib

biellls committed:
Fixed MsSql autocommit bug

jlowin committed:
Update PR template with instructions about JIRA

jlowin committed:
Gracefully fail unit tests when docker-py isn't installed

jlowin committed:
Merge pull request #1456 from airbnb/jlowin-patch-2

[AIRFLOW-34] Update PR template with instructions about JIRA

criccomini committed:
Merge pull request #1458 from biellls/master

[AIRFLOW-27] Fixed MsSql autocommit bug

r39132 committed:
Merge pull request #1430 from whummer/success_recursive

[AIRFLOW-35] Enable UI feature to recursively set success for nested DAG operators

jlowin committed:
Merge pull request #1457 from jlowin/docker-import

[AIRFLOW-38] Gracefully fail unit tests when docker-py isn't installed

mistercrunch committed:
[AIRFLOW-42] Adding logging.debug DagBag loading stats (#1460)

* Adding logging.debug DagBag loading stats

* Linting

* Fix py3

* Tweaks

mistercrunch committed:
Docs tweaks while generating the docs

r39132 committed:
Don't insert dag_runs beyond the min task end_date

criccomini committed:
Merge pull request #1453 from alexvanboxel/feature/AIRFLOW-21-upgrade-gcp-lib

AIRFLOW-21 upgrade GCP client lib

r39132 committed:
Merge pull request #1464 from geeknam/feature/cx_oracle_bulk_insert

[AIRFLOW-50] Add bulk_insert_rows() to OracleHook for more performant inserts.

r39132 committed:
Merge pull request #1466 from r39132/master

[AIRFLOW-39] Don't insert dag_runs beyond the min task end_date

mistercrunch committed:
[AIRFLOW-53] Adding DagBag stats report to CLI's list_dags (#1468)

Adding DagBag stats report to CLI's list_dags

Removing logging call in favor of CLI, on-demend based approach

Addressing Dan's feedback

jlowin committed:
Fix corner case with joining processes/queues (#1473)

If a process places items in a queue and the process is joined before the queue is emptied, it can lead to a deadlock under some circumstances. Closes AIRFLOW-61.

See for example: https://docs.python.org/3/library/multiprocessing.html#all-start-methods ("Joining processes that use queues")
http://stackoverflow.com/questions/31665328/python-3-multiprocessing-queue-deadlock-when-calling-join-before-the-queue-is-em
http://stackoverflow.com/questions/31708646/process-join-and-queue-dont-work-with-large-numbers
http://stackoverflow.com/questions/19071529/python-multiprocessing-125-list-never-finishes

underyx committed:
Add bulk_dump abstract method to DbApiHook (#1471)

jlowin committed:
Add logic to lock DB and avoid race condition

The scheduler can encounter a queued task twice before the
task actually starts to run -- this locks the task and avoids
that condition.

jlowin committed:
[AIRFLOW-52] Fix bottlenecks when working with many tasks

Dag hash function tried (and failed) to hash the list of tasks, then fell back on repr-ing the list, which took forever. Instead, hash tuple(task_dict.keys()). In addition this replaces two slow list comprehensions with much faster hash lookups (using the new task_dict).

aaur0 committed:
Use getfqdn to make sure urls are fully qualified

gethostname only resolves host part while often fully qualified domain names are required.

* Resolves #1437

apapanico committed:
[AIRFLOW-75] Fix bug in S3 config file parsing

whummer committed:
AIRFLOW-77: Enable UI toggle whether to apply 'clear' operation recursively to sub-DAGs or not

Commit:
[AIRFLOW-80] Move example_twitter dag to contrib/example_dags as it requires hive

jlowin committed:
AIRFLOW-52 Warn about overwriting tasks in a DAG

jlowin committed:
Merge pull request #1378 from jlowin/queued-tasks

bolkedebruin committed:
AIRFLOW-92 Avoid unneeded upstream_failed session closes apache/incubator-airflow#1485

jlowin committed:
[AIRFLOW-134] Add PR merge script

mistercrunch committed:
[AIRFLOW-9] Improving docs to meet Apache's standards

r39132 committed:
Fix : Don't treat premature tasks as could_not_run tasks

Commit:
ssl gunicorn support

Commit:
[AIRFLOW-121] Documenting dag doc_md feature

criccomini committed:
Revert "ssl gunicorn support"

This reverts commit e332f63620a5f85e38c4a1a5ac9c9a4a5bfc6035.

artwr committed:
Move presto.execute inside try catch to handle error

This commit fixes an issue where malformed SQL would raise a
DatabaseError outside of the try catch block in the hook. This
should now raise a PrestoException as expected.

r39132 committed:
Merge branch '1490'

aoen committed:
Change default DAG view from tree view to graph view

aoen committed:
[AIRFLOW-112] Change default DAG view from tree view to graph view

aoen committed:
[AIRFLOW-112] no-op README change to close this jira's PR

aoen committed:
Merge branch '1498'

mistercrunch committed:
[AIRFLOW-117] fix links in README.md

Commit:
use targetPartitionSize as the default partition spec

bolkedebruin committed:
AIRFLOW-124 Implement create_dagrun

This adds the create_dagrun function to DAG and the staticmethod
DagRun.find. create_dagrun will create a dagrun including its tasks.

By having taskinstances created at dagrun instantiation time,
deadlocks that were tested for will not take place anymore. Tests
have been adjusted accordingly.

In addition, integrity has been improved by a bugfix to add_task
of the BaseOperator to make sure to always assign a Dag if it is
present to a task.

DagRun.find is a convenience function that returns the DagRuns
for a given dag. It makes sure to have a single place how to
find dagruns.

bolkedebruin committed:
Merge branch '1502'

bolkedebruin committed:
Use incubating instead of incubator in title

Commit:
change TARGET_PARTITION_SIZE to DEFAULT_TARGET_PARTITION_SIZE

aoen committed:
Merge branch '1503'

artwr committed:
[AIRFLOW-109] Fix try catch handling in PrestoHook

This addresses the issue with executing the SQL statement outside of
the try block. In the case of a syntax error in the statement, the
underlying library raises a Databases error which was meant to be
handled (i.e., json parsed) by the catch.

r39132 committed:
Add a version view to display airflow version info

msumit committed:
[AIRFLOW-119] Fix Template not found error and default tags for Qubole operator

criccomini committed:
Merge branch '1493'

Firehed committed:
[AIRFLOW-125] Add file to GCS operator

Adds an operator to upload a file to Google Cloud Storage. Used as follows:
```py
from airflow.contrib.operators.file_to_gcs import FileToGoogleCloudStorageOperator
gcs = FileToGoogleCloudStorageOperator(
        bucket='a-bucket-i-have-access-to-on-gcs',
        dag=dag,
        task_id='upload_stuff',
        google_cloud_storage_conn_id='an-airflow-bigquery-connection',
        src=os.path.join(os.path.dirname(__file__), 'csv/some_file.csv'),
        dst='project/some_file.csv')
```

r39132 committed:
Merge branch '1478'

msumit committed:
AIRFLOW-119: List support for tags in QuboleOperator

Commit:
[AIRFLOW-127] Makes filter_by_owner aware of multi-owner DAG

puneethpotu committed:
Adding Nerdwallet to the list of Currently officially using Airflow:

bolkedebruin committed:
Merge branch 'dag_run'

jlowin committed:
[AIRFLOW-86] Wrap dict.items() in list for Py3 compatibility

Author: jlowin <jlowin@users.noreply.github.com>

Closes #1483 from jlowin/AIRFLOW-86.

criccomini committed:
Merge branch '1507'

criccomini committed:
Merge branch '1508'

r39132 committed:
Increasing License Coverage

sekikn committed:
[AIRFLOW-143] setup_env.sh doesn't leverage cache for downloading minicluster

underyx committed:
Add Kiwi.com as a user to README

aoen committed:
[AIRFLOW-52] 1.7.1 version bump and changelog

aoen committed:
Merge branch '1522'

r39132 committed:
Make enhancements to VersionView

Firehed committed:
[AIRFLOW-148] Use BQ connection project by default in view access

Updates the internals of the BigQuery view access tools to make the source and
destination projects optional, where both values will default to the project
specified in the connection if not explicitly provided.

This is technically a breaking change but I've searched the public codebase and
found no direct calls to this (I'll be updating some internal operators that use
it separately)

Commit:
Updated HiveServer2Hook.to_csv() to add fetch_size

aoen committed:
[AIRFLOW-150] setup.py classifiers dict should be list

r39132 committed:
Merge branch '1520'

criccomini committed:
Merge branch '1505'

criccomini committed:
Merge branch '1524'

aoen committed:
Merge branch '1527'

jlowin committed:
Merge pull request #1515 from jlowin/PR-MERGE

mred-cmd committed:
[AIRFLOW-153] fixed minor typos in help message

criccomini committed:
Merge pull request #1533 from markreddy/AIRFLOW-153-fix-typos

jlowin committed:
[AIRFLOW-157] Make PR tool Py3-compat; add JIRA command

- Adds Python3 compatibility (filter objects can't be indexed)
- Adds JIRA command to close issues without merging a PR
- Adds general usability fixes and starts cleaning up code

mistercrunch committed:
Pointing setup.py to then new repo

mistercrunch committed:
1.7.1.1

aoen committed:
Bump version to unblock pypi release

mreid-moz committed:
docfix: Fix a couple of minor typos.

aoen committed:
Merge branch '1526'

msumit committed:
AIRFLOW-167: Add dag_state option in cli

Firehed committed:
[AIRFLOW-170] Add missing @apply_defaults

criccomini committed:
Merge pull request #1542 from Firehed/apply_defaults

jlowin committed:
[AIRFLOW-175] Run git-reset before checkout in PR tool

If the user made any changes, git checkout will fail because the
changes would be overwritten. Running git reset blows the changes away.

jlowin committed:
Merge pull request #1534 from jlowin/pr-tool-2

Commit:
AIRFLOW-45: Support Hidden Airflow Variables

jlowin committed:
[AIRFLOW-176] Improve PR Tool JIRA workflow

- Fix crash when non-integer IDs are passed
- Improve workflow by always asking user if they
  want to resolve another issue before exiting

criccomini committed:
Merge pull request #1530 from mattuuh7/hidden-fields

Commit:
[AIRFLOW-178] Fix bug so that zip file is detected in DAG folder

jlowin committed:
Merge pull request #1544 from jlowin/pr-tool-3

jlowin committed:
[AIRFLOW-176] remove unused formatting key

criccomini committed:
Merge pull request #1545 from jgao54/zip-bug-fix

jlowin committed:
Merge pull request #1546 from jlowin/pr-tool-4

criccomini committed:
Merge pull request #1541 from msumit/AIRFLOW-167

artwr committed:
[AIRFLOW-180] Fix timeout behavior for sensors

In the previous state of the code, datetime.now was compared to
started_at and seconds was pulled out. It turns out that the seconds
attribute of a timedelta has a maximum of 86400 and the rolls up to 1 day.
The unintended consequence is that timeout larger than 86400 are
ignored, with sensors running forever.

To fix this we use the total_seconds method to get at the real
timedelta in seconds.

bolkedebruin committed:
AIRFLOW-168 Correct evaluation of @once schedule

If the schedule @once was used with a start_date two dagruns
would be created as next_run_date would be none and compared
against dag.start_date. This patch fixes that by returning
immediately if a dagrun has already occured with an @once
schedule.

bolkedebruin committed:
AIRFLOW-181 Fix failing unpacking of hadoop by redownloading

curl compares timestamps, but if the file is corrupt this can
result in hadoop tars that are never updated. This adds a retry
without using the cache.

sayap committed:
[AIRFLOW-183] Fetch log from remote when worker returns 4xx/5xx response

Dear Airflow Maintainers,

Please accept this PR that addresses the following issues:

- https://issues.apache.org/jira/browse/AIRFLOW-183

This is mainly to make the behavior consistent when some log files have
been deleted from the log folder. Without the change, the remote s3/gcs
fallback will only trigger if the task ran on the local worker.

Author: Yap Sok Ann <sokann@gmail.com>

Closes #1551 from sayap/remote-log-remote-worker.

jlowin committed:
[AIRFLOW-185] Handle empty versions list

Commit:
[AIRFLOW-179] DbApiHook string serialization fails when string contains non-ASCII characters

Dear Airflow Maintainers,

Please accept this PR that addresses the following issues:
- https://issues.apache.org/jira/browse/AIRFLOW-179

In addition to correctly serializing non-ASCII characters the literal transformation also corrects an issue with escaping single quotes (').

Note it was my intention to add another unit test to `test_hive_to_mysql` in `tests/core.py` however on inspection the indentations of the various methods seemed wrong, methods are nested and it's not apparent what class they refer to. Additionally it seems a number of the test cases aren't related to the corresponding class.

For testing purposes I simply ran a pipeline which previously failed with the following exception,

    [2016-05-26 22:03:39,256] {models.py:1286} ERROR - 'ascii' codec can't decode byte 0xc3 in position 230: ordinal not in range(128)
    Traceback (most recent call last):
      File "/usr/local/lib/python2.7/dist-packages/airflow/models.py", line 1245, in run
    result = task_copy.execute(context=context)
      File "/usr/local/lib/python2.7/dist-packages/airflow/operators/hive_to_mysql.py", line 88, in execute
    mysql.insert_rows(table=self.mysql_table, rows=results)
      File "/usr/local/lib/python2.7/dist-packages/airflow/hooks/dbapi_hook.py", line 176, in insert_rows
    l.append(self._serialize_cell(cell))
      File "/usr/local/lib/python2.7/dist-packages/airflow/hooks/dbapi_hook.py", line 196, in _serialize_cell
    return "'" + str(cell).replace("'", "''") + "'"
      File "/usr/local/lib/python2.7/dist-packages/future/types/newstr.py", line 102, in __new__
    return super(newstr, cls).__new__(cls, value)
    UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 230: ordinal not in range(128)

and verified with the presence of the fix that the task succeeded and the resulting output was correct. Note currently from grokking the code base it seems that only `MySqlHook` objects call the the `insert_rows` method.

Author: John Bodley <john.bodley@airbnb.com>

Closes #1550 from johnbodley/dbapi_hook_serialization.

artwr committed:
Revert "[AIRFLOW-179] DbApiHook string serialization fails when string contains non-ASCII characters"

This reverts commit 87b4b8fa19cb660317198d74f6d51fdde0a7e067.

Reverting as the method used in the dbapi hook is actually package
specific to MySQLdb and would break the sqlite and mssql hooks.

bolkedebruin committed:
Merge branch 'AIRFLOW-168'

bolkedebruin committed:
AIRFLOW-190 Add codecov and remove download count

bolkedebruin committed:
Do not add comments to pull requests

r39132 committed:
[AIRFLOW-195] : Add toggle support to subdag clearing in the CLI

sekikn committed:
[AIRFLOW-25] Configuration for Celery always required

Dear Airflow Maintainers,

Please accept this PR that addresses the following issues:
- AIRFLOW-25

For now, if airflow.cfg has no [celery] section, all subcommands fail.
This patch adds the default values for Celery-related properties
as well as existing 'default_queue' and 'flower_port',
so as to make all subcommands work and suppress "not found in config"
warnings even if [celery] section is omitted.

Author: Kengo Seki <sekikn@apache.org>

Closes #1558 from sekikn/AIRFLOW-25.

alexvanboxel committed:
[AIRFLOW-23] Support for Google Cloud DataProc

Dear Airflow Maintainers,

Please accept this PR that addresses the following issues:
- AIRFLOW-23

Spark, Hadoop, PySpart, SparkSQL, Pig an Hive support of DataProc cluster

Author: Alex Van Boxel <alex@vanboxel.be>

Closes #1532 from alexvanboxel/AIRFLOW-23.

bolkedebruin committed:
Optimize and refactor process_dag

This patch addresses the following issues:
get_active_runs was a getter that was also updating
to the database. This patch refactors get_active_runs
into two different functions that are part of DagRun.
update_state update state of the dagrun based on the
taskinstances of the dagrun. verify_integrity checks
and updates the dag run based on if the dag contains
new or missing tasks.

Deadlock detection has been refactored to ensure that
database does not get hit twice, in some circumstances
this can reduce the time spent by 50%.

process_dag has been refactored to use the functions
of DagRun reducing complexity and reducing pressure on the
database. In addition locking is now properly working
under the assumption that the heartrate is longer
than the time process_dag spends.

Two new TaskInstance states have been introduced. "REMOVED"
and "SCHEDULED". REMOVED will be set when taskinstances
are encountered that do no exist anymore in the DAG.
This happens when a DAG is changed (ie. a new version).
The "REMOVED" state exists for lineage purposes.
"SCHEDULED" is used when a Task that did not have a state
before is sent to the executor. It is used by both the
scheduler and backfills. This state almost removes the
race condition that exists if using multiple schedulers:
due to the fact UP_FOR_RETRY is being managed by the
TaskInstance (I think that is the wrong place) is still
exists for that state.

r39132 committed:
Merge branch '1557'

junwei-wang committed:
[AIRFLOW-196] Fix bug that exception is not handled in HttpSensor

Dear Airflow Maintainers,

Please accept this PR that addresses the following issues:
- [*AIRFLOW-196*](https://issues.apache.org/jira/browse/AIRFLOW-196)

If exception happens in poke function in HttpSensor, it is not
well handled that make the sensor finish successfully, which is
incorrect obviously.

Author: Junwei Wang <i.junwei.wang@gmail.com>

Closes #1561 from junwei-wang/master.

austyantsev-mail committed:
AIRFLOW-202: Fixes stray print line

austyantsev-mail committed:
[AIRFLOW-201] Fix for HiveMetastoreHook + kerberos

Dear Airflow Maintainers,

Please accept this PR that addresses the following issues:
- https://issues.apache.org/jira/browse/AIRFLOW-201

Author: Alexey Ustyantsev <a.ustyantsev@corp.mail.ru>

Closes #1563 from hudbrog/airflow_201.

bolkedebruin committed:
Merge pull request #1564 from hudbrog/airflow_202

jlowin committed:
[AIRFLOW-187] Improve PR tool UX

- Add command to setup_git_remotes
- Add guard for running PR tool in a branch with uncommitted changes
- Add option to edit squash commit message
- Message styling for clarity
- Improved error messages for PR Tool problems

bolkedebruin committed:
Merge branch 'process_dag'

msumit committed:
[AIRFLOW-155] Documentation of Qubole Operator

Dear Airflow Maintainers,

Please accept this PR that addresses the following issues:
- *https://issues.apache.org/jira/browse/AIRFLOW-155*

Thanks,
Sumit

Author: Sumit Maheshwari <sumitm@qubole.com>

Closes #1560 from msumit/AIRFLOW-155.

jlowin committed:
Merge pull request #1565 from jlowin/pr-tool-5

jlowin committed:
[AIRFLOW-207] Improve JIRA auth workflow

bolkedebruin committed:
[AIRFLOW-209] Add scheduler tests and improve lineage handling

This patch adds schedule_dag and process_dag unittests. It also
fixes some minor bugs that were caught by these tests. Some
small changes for readability.

bolkedebruin committed:
Merge branch 'master' of https://git-wip-us.apache.org/repos/asf/incubator-airflow

jlowin committed:
[AIRFLOW-187] Improve PR Tool UX

Improved styling for PR tool prompts

jlowin committed:
Merge pull request #1567 from jlowin/pr-tool-7

Commit:
[AIRFLOW-114] Sort plugins dropdown

Closes #1499 from varantz/sorting_plugins_in_dropdown.

itajaja committed:
[AIRFLOW-64] Add note about relative DAGS_FOLDER

Closes #1474 from itajaja/patch-1.

jlowin committed:
[AIRFLOW-211] Fix JIRA "resolve" vs "close" behavior

Closes #1571 from jlowin/pr-tool-8.

dantino committed:
[AIRFLOW-206] Always load local log files if they exist

jlowin committed:
[AIRFLOW-206] Add commit to close PR

Closes #1566
GitHub requires the above string to properly mark the PR as closed

jlowin committed:
[AIRFLOW-213] Add "Closes #X" phrase to commit messages

bolkedebruin committed:
[AIRFLOW-214] Fix occasion of detached taskinstance

For some reason occasionely taskinstanced could become
detached from the database session. Now it uses a fresh session
to ensure the taskinstances stay attached.

Commit:
[AIRFLOW-131] Make XCom.clear more selective

XCOMs for a task were getting cleared on every run, no matter what.
Selectively clear only when the task is actually going to be run.

Closes #1570 from johnnason/AIRFLOW-131

bolkedebruin committed:
Merge remote-tracking branch 'apache/master'

bolkedebruin committed:
[AIRFLOW-68] Align start_date with the schedule_interval

This particular issue arises because of an alignment issue between
start_date and schedule_interval. This can only happen with cron-based
schedule_intervals that describe absolute points in time (like “1am”) as
opposed to time deltas (like “every hour”)

In the past (and in the docs) we have simply said that users must make
sure the two params agree. But this is counter intuitive. As in these
cases, start_date is sort of like telling the scheduler to
“start paying attention” as opposed to “this is my first execution date”.

This patch changes the behavior of the scheduler. The next run date of
the dag will be treated as "start_date + interval" unless the start_date
is on the (previous) interval in which case the start_date will be the
next run date.

jlowin committed:
Merge pull request #1573 from jlowin/test-merge-pr

mistercrunch committed:
Merge pull request #1569 from mistercrunch/docs

aoen committed:
[AIRFLOW-218] Added option to enable webserver gunicorn access/err logs

Closes #1577 from aoen/ddavydov/better_http_response_logging

Added an option to enable gunicorn access/error logs.
The default config will now log webserver errors/accesses to stderr.

Also made the 404 page hostname text default page color instead of
white, since white text is pretty hard to see against a white
background.

Commit:
[AIRFLOW-223] Make parametrable the IP on which Flower binds to

sekikn committed:
[AIRFLOW-142] setup_env.sh doesn't download hive tarball if hdp is specified as distro

Closes #1518 from sekikn/AIRFLOW-142

bolkedebruin committed:
Merge branch 'align_startdate'

bolkedebruin committed:
[AIRFLOW-224] Collect orphaned tasks and reschedule them

Tasks can get orphaned if the scheduler is killed in the middle
of processing the tasks or if the MQ queue is cleared without
a worker having picked these up. Now tasks do not get set
to a scheduled state anymore if they have not been sent to the
executor yet. Next to that a garbage collector scans the executor
for tasks not being present and reschedules those if needed.

mistercrunch committed:
[AIRFLOW-230] [HiveServer2Hook] adding multi statements support

Changing the library from pyhive to impyla broke the behavior where multiple statements, including statements that don't return results were previously supported and aren't anymore. impyla raises an exception if any of the statements doesn't return result.

We have tasks that run multiple statements including DDL and want to run them atomically.

Closes #1583 from mistercrunch/hooks_hive_presto

[AIRFLOW-230] [HiveServer2Hook] adding multi statements support

mistercrunch committed:
[AIRFLOW-238] Make compatible with flask-admin 1.4.1

The new flask-admin==1.4.1 release on 2016-06-13 breaks the Airflow
release currently in Pypi (1.7.1.2). This fixes the edge case triggered
by this new release.

* Closes #1588 on github

rfroetscher committed:
[AIRFLOW-171] Add upgrade notes on email and S3 to 1.7.1.2

Closes #1587 from rfroetscher/upgrading_readme

jwi078 committed:
[AIRFLOW-216] Add Sqoop Hook and Operator

This patch adds a Sqoop hook and operator that implements Sqoop import.
The hook is a wrapper around the sqoop 1 binary.

* Closes #1576 from jwi078/sqoop_operator

mistercrunch committed:
[AIRFLOW-231] Do not eval user input in PrestoHook

Running `eval` represent a security threat as the interpreter can be
hijacked by the service returning the string getting "evaled", in this
case Presto. It turns out the code I'm changing here was written a long
time ago and misguided, casting a python object to a string and then
evaling it as a useless round trip.

Closes #1584 from mistercrunch/security

sekikn committed:
[AIRFLOW-222] Show duration of task instances in ui

At the moment mousing over a task instance in the Airflow tree view
only shows the duration for completed task, and shows "Duration: null"
for running tasks.

Instead the UI should show the current task instance's duration
for running tasks. This patch addresses this.

Closes #1589 from sekikn/AIRFLOW-222

aoen committed:
[AIRFLOW-241] Add testing done section to PR template

criccomini committed:
Merge pull request #1592 from aoen/ddavydov/add_testing_done_section_to_pr_template

sekikn committed:
[AIRFLOW-225] Better units for task duration graph

Right now the job duration window defaults to hours, which for short lived tasks
results in numbers out to five decimals. This patch adjusts the scale of the Y-axis
in accordance with the maximum value of the durations to be shown.

ldct committed:
[AIRFLOW-239] Fix tests indentation

Closes #1591 from zodiac/deindent-tests

aoen committed:
Merge branch '1582'

trixpan committed:
[AIRFLOW-173] Initial implementation of FileSensor

Implement a simple FileSensor to detect the creation of files and folder
in a given filesystem (a la HDFSSensor)

Closes #1543 from trixpan/AIRFLOW-173

jlowin committed:
[AIRFLOW-31] Use standard imports for hooks/operators

bolkedebruin committed:
Merge branch 'master' of https://git-wip-us.apache.org/repos/asf/incubator-airflow

criccomini committed:
Merge pull request #1595 from sekikn/AIRFLOW-225

jlowin committed:
Add Python 3 compatibility fix

In Python 3, errors don’t have a `message` attribute

Syeoryn committed:
Add Postmates to Airflow users list

Closes #1599 from Syeoryn/master

Add Postmates to Airflow users list

jlowin committed:
Merge pull request #1272 from jlowin/standard-imports

bolkedebruin committed:
[AIRFLOW-256] Fix test_scheduler_reschedule heartrate

test_scheduler_reschedule runs two schedulerjob quite
fast after one another this sometimes is faster than
the heartrate allows and thus the tasks will not get
rescheduled and the test will fail. Fixed by setting
heartrate to 0.

bolkedebruin committed:
Merge remote-tracking branch 'apache/master'

mistercrunch committed:
[AIRFLOW-234] make task that aren't `running` self-terminate

Closes #1585 from mistercrunch/undeads

bolkedebruin committed:
[AIRFLOW-6] Remove dependency on Highcharts

Highcharts' license is not compatible with the Apache 2.0
license. This patch removes Highcharts in favor of d3,
however some charts are not supported anymore.

* This brings Maxime Beauchemin's work to master

bolkedebruin committed:
Merge branch 'highcharts_to_d3'

jlowin committed:
[AIRFLOW-31] Add zope dependency

Closes #1608 from jlowin/standard-imports-2.
Also closes AIRFLOW-257.

ajayyadava committed:
[AIRFLOW-262] Simplify commands in MANIFEST.in

artwr committed:
Merge pull request #1547 from artwr/artwr_fix_sensor_timeout

sekikn committed:
[AIRFLOW-252] Raise Sqlite exceptions when deleting tasks instance in WebUI

If users who use SQLite as backend try to delete a task via browser, it fails with an exception.
Though this is a bug on Flask-Admin's side basically, this patch provides a workaround for it.

criccomini committed:
Merge pull request #1612 from ajayyadava/262

artwr committed:
[AIRFLOW-263] Remove temp backtick file

Closes #1613 from artwr/artwr_remove_backtick_file

[AIRFLOW-263] Remove temp backtick file

This removes the file erroneously introduced by the highcharts
refactor.

Solves AIRFLOW-263

ajayyadava committed:
[AIRFLOW-248] Add Apache license header to all files

- Added Apache license header for files with extension (.service, .in, .mako, .properties, .ini, .sh, .ldif, .coveragerc, .cfg, .yml, .conf, .sql, .css, .js, .html, .xml.
- Added/Replaced shebang on all .sh files with portable version - #!/usr/bin/env bash.
- Skipped third party css and js files. Skipped all minified js files as well.

Closes #1598 from ajayyadava/248

criccomini committed:
Merge pull request #1609 from sekikn/AIRFLOW-252

alexvanboxel committed:
[AIRFLOW-162] Allow variable to be accessible into templates

Closes #1540 from alexvanboxel/AIRFLOW-162

AIRFLOW-162 Allow variable to be accessible into templates

paulbramsen committed:
[AIRFLOW-244] Modify hive operator to inject analysis data

Testing Done:
Test dags were run as backfills on an Airbnb Airflow dev box.

This PR exposes task/dag id/run data through the HiveOperator for
ingestion by performance analysis tools like Dr. Elephant.

Closes #1607 from paulbramsen/paulbramsen/modify_hive_operator_to_inject_analysis_data

ldct committed:
Start reader thread

ldct committed:
whitespace

bolkedebruin committed:
[AIRFLOW-275] Update contributing guidelines

The current guidelines do not reflect the move to Apache. This
updates the links and update the PR template.

Commit:
[AIRFLOW-273] Create an svg version of the airflow logo.

Closes #1619 from gwax/svg_logo

ldct committed:
Move shared dagbag to shared.py, create on import

ldct committed:
Move refresh_dags

ldct committed:
change import path

plypaul commented on airflow/utils/timeout.py:
```diff
@@ -39,8 +40,9 @@ def handle_timeout(self, signum, frame):
         try:
-            signal.signal(signal.SIGALRM, self.handle_timeout)
-            signal.alarm(self.seconds)
+            if isinstance(threading.current_thread(), threading._MainThread):
```
If a caller tries to use this outside of the main thread, should we throw an exception?

ldct replied:
We'd have to make `process_file` stop using timeout if it's not on the main thread (https://github.com/apache/incubator-airflow/blob/master/airflow/models.py#L252)

plypaul replied:
Well as it is now, this passes silently, right?

ldct replied:
Yeah, no signal handler is installed on the background thread and long loading dags are not killed

plypaul commented on airflow/models.py:
```diff
@@ -180,6 +178,14 @@ def size(self):
+            example_dag_folder = os.path.join(
+                os.path.dirname(__file__),
+                'example_dags')
+            self.collect_dags(example_dag_folder)
```
Is `collect_dags` thread safe?

Commit:
[AIRFLOW-274] Add XCom functionality to GoogleCloudStorageDownloadOperator

Updated GoogleCloudStorageDownloadOperator so that it has the option to
store the downloaded file's contents in an XCom instead of saving to
disk. It now also takes filename as an optional argument instead of a
required argument, so it is not necessary to save to disk if you do not
need to.

Closes #1618 from illop/risk_analytics

ldct committed:
Add license to shared.py

ldct committed:
kick

Commit:
[AIRFLOW-280] clean up tmp druid table no matter if an ingestion job succeeds or not

Closes #1624 from hongbozeng/cleanup_druid

clean up tmp druid table no matter if the ingestion job success or not

bilevn committed:
[AIRFLOW-278] Support utf-8 ecoding for SQL

Support utf-8 encoding for SQL queries, needed for Python 2
users who have unicode strings inside the queries

Closes #1622 from biln/master

Commit:
[AIRFLOW-283] Make store_to_xcom_key a templated field in GoogleCloudStorageDownloadOperator

Closes #1628 from illop/gcs_download_operator

criccomini committed:
[AIRFLOW-200] Make hook/operator imports lazy, and print proper exceptions

criccomini committed:
[AIRFLOW-285] Use Airflow 2.0 style imports for all remaining hooks/operators

bolkedebruin committed:
Merge branch 'contributing'

dsjl committed:
[AIRFLOW-40] Add LDAP group filtering feature.

It is now possible to filter over LDAP group (in the web
interface) when using the LDAP authentication backend.
Note that this feature requires the "memberOf"
overlay to be configured on the LDAP server.

Closes #1479 from dsjl/AIRFLOW-40

lukembis committed:
[AIRFLOW-281] Add port to mssql_hook

Closes #1626 from lukem-ow/master

criccomini committed:
Merge pull request #1586 from criccomini/AIRFLOW-200

reconditesea committed:
[ARFLOW-255] Check dagrun timeout when comparing active runs

Timedout DagRuns and max active runs reached would be set to failed.
Therefor, concurrency limits could be reached prematurely.

Closes #1604 from reconditesea/klin-dagrun-timeout

withnale committed:
[AIRFLOW-189] Highlighting of Parent/Child nodes in Graphs

Closes #1554 from withnale/graph_highlight

sekikn committed:
[AIRFLOW-246] Improve dag_stats endpoint query

For now, accessing /dag_stats can take a relatively long time
(e.g. over 20 seconds with less than a million rows on some environment).
This patch replaces multiple LEFT OUTER JOINs with INNER JOINs and UNION ALL
and improves that process by making it 3-5x faster.

Closes #1610 from sekikn/AIRFLOW-246

mistercrunch committed:
Add an Apache Incubator Disclaimer and mocking modules

Closes #1634 from mistercrunch/mock_docs

Adding an Apache Incubator Disclaimer and mocking modules

ldct committed:
[AIRFLOW-243] Create NamedHivePartitionSensor

Closes #1593 from zodiac/create-NamedHivePartitionSensor

skudriashev committed:
[AIRFLOW-286] Improve FTPHook to implement context manager interface

Closes #1632 from skudriashev/airflow-286

msumit committed:
[AIRFLOW-296] template_ext is being treated as a string rather than a tuple in qubole operator

Closes #1638 from msumit/AIRFLOW-296

sekikn committed:
[AIRFLOW-269] Add some unit tests for PostgreSQL

Closes #1616 from sekikn/AIRFLOW-269

jlowin committed:
[AIRFLOW-100] Add execution_date_fn to ExternalTaskSensor

Currently, ExternalTaskSensor only supports querying execution_dates
that are either the same as the ExternalTaskSensor's execution_date
or a fixed interval from that date (using `execution_delta`). This
adds the ability to provide a fn (`execution_date_fn`) that accepts
the current execution_date and can return any desired date for
querying. This is much more flexible. For example, it could
supply the last date of the previous month.

aoen committed:
[AIRFLOW-291] Add index for state in TI table

Closes #1635 from aoen/ddavydov/add_index_to_task_instance_state

jlowin committed:
[AIRFLOW-301] Fix broken unit test

This unit tests always fails on the last day of the month, since it
tries to access a nonexistent day (like June 31st).

jlowin committed:
[AIRFLOW-282] Remove PR Tool logic that depends on version formatting

Closes #1625 from jlowin/pr-tool

jlowin committed:
Add license to migration file

rfroetscher committed:
[AIRFLOW-247] Add EMR hook, operators and sensors. Add AWS base hook

Closes #1630 from rfroetscher/emr

jlowin committed:
Merge pull request #1641 from jlowin/external-task

Commit:
[AIRFLOW-284] HiveServer2Hook fix for cursor scope for get_results

Dear Airflow Maintainers,

Please accept this PR that addresses the following issues:
- *(replace with a link to AIRFLOW-X)*

Detail:
The function `get_results` implemented in the `HiveServer2Hook` does not execute multiple commands passed to it in a list, in the singular cursor scope. This has caused SQL statements that depend on the execution of `add jar` and `set` commands to fail as they are being executed in different cursor scopes which are not persistent.
The code has been updated to have the cursor object persistent.

Testing Done:

Reminders for contributors:
* Your PR's title must reference an issue on [Airflow's JIRA](https://issues.apache.org/jira/browse/AIRFLOW/). For example, a PR called "[AIRFLOW-1] My Amazing PR" would close JIRA issue #1. Please open a new issue if required!
* Please squash your commits when possible and follow the [7 rules of good Git commits](http://chris.beams.io/posts/git-commit/#seven-rules).

mistercrunch

Closes #1629 from sherwian/hivehook_fx

mistercrunch committed:
[AIRFLOW-298] fix incubator diclaimer in docs

Closes #1640 from mistercrunch/disclaimer_tweaks

[AIRFLOW-298] fix incubator diclaimer in docs

mtagle committed:
[AIRFLOW-314] Fix BigQuery cursor run_table_upsert method

Closes #1652 from mtagle/fix_bq_table_upsert

By default, bigquery will only return 50 tables when you ask for a list
of all the tables in a datatset. If you are trying to upsert a table
that exists, but you have more than 50 tables, the run_table_upsert
method may conclude that the table doesn't exist, and try to insert it,
and bigquery will error saying that the table does exist.

This fix checks if the response has pagination data, and looks at all
the pages, rather than just the first one, to see if the table exists.

Firehed committed:
[AIRFLOW-308] Add link to refresh DAG within DAG view header

Closes #1646 from Firehed/refresh_on_dag

alexvanboxel committed:
[AIRFLOW-24] DataFlow Java Operator

Closes #1648 from alexvanboxel/AIRFLOW-24

skudriashev committed:
[AIRFLOW-311] Fix wrong path in CONTRIBUTING.md

Closes #1650 from skudriashev/airflow-311

skudriashev committed:
[AIRFLOW-313] Fix code style for sqoop_hook.py

Closes #1651 from skudriashev/airflow-313

oza committed:
[AIRFLOW-307] Rename __neq__ to __ne__ python magic method.

Closes #1649 from oza/AIRFLOW-307

oza committed:
[AIRFLOW-309] Add requirements of develop dependencies to docs

Closes #1647 from oza/AIRFLOW-309

Commit:
[AIRFLOW-307] There is no __neq__ python magic method.

Closes #1645 from TheSAS/patch-1

jlowin committed:
[AIRFLOW-31][AIRFLOW-200] Add note to updating.md

AIRFLOW-31 and AIRFLOW-200 deprecated the old important mechanism and should be noted in UPDATING.md

Closes #1643 from jlowin/patch-1

Commit:
[AIRFLOW-297] support exponential backoff option for retry delay

Closes #1639 from jgao54/support-retry-backoff

ldct committed:
use settrace to timeout on child thread, refactor dagbag code

ldct committed:
avoid parsing dagbag twice on list_dags

ldct committed:
print dagbag stats in background thread

ldct committed:
only report on new dags

ldct committed:
Reduce frequency of time checks

ldct committed:
Fix old reference to dagbag
